open-kilda-5264 GUI: Add ability to reroute YFlow via UI

IvanChupin opened the pull request:
### This PR should be merged after the following one: (https://github.com/telstra/open-kilda/pull/5318)

UI:
- add y_flow_id to the flowDetail object that is stored in the localStorage.
- retrieve this y_flow_id when it is not null for particular flow, and call appropriate API for y-flow rerouting when needed.
- propagated the rerouting error from the northbound API.
- changed button name "Reroute Flow"->"Reroute Y-Flow" for Y-Flow subflows.
- added y-flow column to the flow-datables component.
- added Y-Flow ID list item to the Flow Details Page, flow details section for the subflow case.
- remove edit button for the Flow details page of the subflow case.

BFF:
  - add additional Controller,Service for Y-flow.

closes: #5264 






In order to test the ability to re-route Y-Flow we need to follow the next steps:

1) Create an Y-Flow using [y-flow-controller-v-2], bellow is the possible request body for creating Y-Flow.
```
{
  "y_flow_id": "yflow8-9-3",
  "shared_endpoint": {
    "port_number": 101,
    "switch_id": "8"
  },
  "maximum_bandwidth": 100,
  "path_computation_strategy": "cost",
  "sub_flows": [
    {
      "flow_id": "subflow8-9",
      "endpoint": {
        "detect_connected_devices": {
          "arp": true,
          "lldp": true
        },
        "inner_vlan_id": 108,
        "port_number": 103,
        "switch_id": "9",
        "vlan_id": 109
      },
      "shared_endpoint": {
        "inner_vlan_id": 110,
        "vlan_id": 111
      }
    },
{
      "flow_id": "subflow8-3",
      "endpoint": {
        "detect_connected_devices": {
          "arp": true,
          "lldp": true
        },
        "inner_vlan_id": 118,
        "port_number": 113,
        "switch_id": "3",
        "vlan_id": 120
      },
      "shared_endpoint": {
        "inner_vlan_id": 111,
        "vlan_id": 112
      }
    }
  ]
}
```

2) After successful creation of Y-Flow we need to reroute the flow, but just calling reroute 
API ([/v2/y-flows/{y_flow_id}/reroute]), it only gives us the error message with the following content:
```
{
  "correlation_id": "54234e83-0e52-4287-b7bc-55f14a512018 : CHUPIN_REROUTE",
  "timestamp": 1689864780401,
  "error-type": "Internal service error",
  "error-message": "Could not reroute y-flow",
  "error-description": "Reroute is unsuccessful. Couldn't find new path(s)"
}
```
This happens because of the existed paths are already best routed according to the internal logic of the flowhs-topology, despite the fact that an error message obviously misleading. **However we have already had a fix for this problem in the following PR**: 
https://github.com/telstra/open-kilda/pull/5318.
To trick this behaviour we need to change the cost of the existing links in our two paths, to do so as a one easy option we need to follow this link: (http://vm:2480/studio/index.html#/),  enter the credentials, 
follow the "GRAPH" tab and run the query:
<img width="1292" alt="image" src="https://github.com/telstra/open-kilda/assets/9297385/ce7e542b-d328-4269-a2f2-a668c8bf5e60">

change the cost for links between switches 8-3 and switches 8-9.
<img width="1210" alt="image" src="https://github.com/telstra/open-kilda/assets/9297385/e64ab065-f57e-4d3a-a9d6-4267adace521">
Setting the cost to 7000 for each link would be enough:
<img width="796" alt="image" src="https://github.com/telstra/open-kilda/assets/9297385/5e7a1164-b976-4a26-8fc5-b7428214148f">.

Then go to the UI
<img width="2289" alt="image" src="https://github.com/telstra/open-kilda/assets/9297385/4574b39b-b8b7-4e86-94c0-8fa4e57e45b5">
<img width="2559" alt="image" src="https://github.com/telstra/open-kilda/assets/9297385/f066930e-f054-4d7a-a599-02c369fcfeff">
Pressing the button should reroute all the subflows for the desired Y-Flow's subflow.


**Here are screenshots with the UI changes:**
<img width="2296" alt="image" src="https://github.com/telstra/open-kilda/assets/9297385/e6e79831-51e7-407f-9f4c-36a1f55bd1b2">
<img width="2294" alt="image" src="https://github.com/telstra/open-kilda/assets/9297385/8eee6116-fdb8-4d37-927b-ddb532922b8e">

1. We do not have the edit button for the Y-Flow subflows anymore.
2. Added the Y-Flow ID field with the value, for the Y-Flow subflows(there is no this field for regular flow)
3. Changed the label for the reroute flow button.
4. Added Y-Flow ID column with a search feature. 



### This PR should be merged after the following one: (https://github.com/telstra/open-kilda/pull/5318)

pablomuri commented on src-gui/src/main/java/org/openkilda/integration/service/FlowsIntegrationService.java:
```diff
@@ -278,7 +304,7 @@ public FlowV2 createFlow(FlowV2 flow) {
      *
      * @param flowId the flow id
-     * @param flow the flow
+     * @param flow   the flow
```
additional space

pablomuri commented on src-gui/src/main/java/org/openkilda/integration/service/FlowsIntegrationService.java:
```diff
@@ -350,11 +376,11 @@ public String resyncFlow(String flowId) {
      * Flow ping.
      *
-     * @param flow the flow
+     * @param flow   the flow
```
same

IvanChupin replied:
thanks, fixed

pablomuri commented on src-gui/src/main/java/org/openkilda/model/FlowInfo.java:
```diff
@@ -40,6 +40,9 @@ public class FlowInfo implements Serializable {
     @JsonProperty("flowid")
     private String flowid;
 
+    @JsonProperty("y_flow_id")
+    private String yFlowId;
+
```
I think this class can be simplified with the @JsonNaming(value = SnakeCaseStrategy.class) annotation, can't it?

IvanChupin replied:
You are right, but for our case, this is not the best approach. In this class, we have snake case for some fields, kebab case for other fields. Moreover, we have mixes of two cases in one field. How about this:
    @JsonProperty("target-path-computation_strategy")
    private String targetPathComputationStrategy;
    
    
Although it is a good idea to use this class-level annotation @JsonNaming(value = SnakeCaseStrategy.class), I would leave it as it is for now and create an additional task to fix all these kinds of things.

Because you understand, if we want to unify everything in this class, then I have to prepare for this unification in the UI part, which could cause many more additional changes.

pablomuri replied:
An intermediate option could be to set the SnakeCaseStrategy and let the special cases be manually specified. Right now is not necessary to unification all the UI code but I agree we can do it later.

IvanChupin replied:
fixed

pablomuri reviewed (changes requested):
Have you implemented a notice to the user saying that both subflows will be rerouted?

dmitrii-beliakov reviewed (approved):
👍  The PR description is perfect!

dmitrii-beliakov commented on src-gui/src/main/java/org/openkilda/controller/YFlowController.java:
```diff
@@ -0,0 +1,60 @@
+    private ActivityLogger activityLogger;
+
+    /**
+     * Re route the y-flow and returns shared path and sub-flow paths with all nodes/switches exists in
```
```suggestion
     * Reroutes the y-flow and returns the shared path and sub-flow paths with all nodes/switches that exist in the
```

IvanChupin replied:
fixed

dmitrii-beliakov commented on src-gui/src/main/java/org/openkilda/controller/YFlowController.java:
```diff
@@ -0,0 +1,60 @@
+     * Re route the y-flow and returns shared path and sub-flow paths with all nodes/switches exists in
+     * provided y-flow.
+     *
+     * @param yFlowId id of reroute requested.
```
```suggestion
     * @param yFlowId ID of the y-flow to be rerouted.
```

dmitrii-beliakov commented on src-gui/src/main/java/org/openkilda/controller/YFlowController.java:
```diff
@@ -0,0 +1,60 @@
+     * provided y-flow.
+     *
+     * @param yFlowId id of reroute requested.
+     * @return reroute flow of new flow path with all nodes/switches exist
```
```suggestion
     * @return the result of the rerouting operation.
```

IvanChupin replied:
fixed

dmitrii-beliakov commented on src-gui/src/main/java/org/openkilda/integration/model/FlowV2.java:
```diff
@@ -37,6 +36,9 @@ public class FlowV2 {
     @JsonProperty("flow_id")
     private String id;
 
+    @JsonProperty("y_flow_id")
```
Consider to use a class level annotation for the snake case.

dmitrii-beliakov commented on src-gui/src/main/java/org/openkilda/integration/service/FlowsIntegrationService.java:
```diff
@@ -200,6 +201,33 @@ public FlowPath rerouteFlow(String flowId) {
+     * @param yFlowId y-flow id
+     * @return YFlowRerouteResult
+     */
+    public YFlowRerouteResult rerouteYFlow(String yFlowId) {
```
do think require `yFlowId` to be non-null will be useful here?

IvanChupin replied:
fixed

dmitrii-beliakov commented on src-gui/src/main/java/org/openkilda/model/YFlowRerouteResult.java:
```diff
@@ -0,0 +1,66 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
```
```suggestion
@JsonInclude(JsonInclude.Include.NON_EMPTY)
@FieldDefaults(level = AccessLevel.PRIVATE)
```

IvanChupin replied:
fixed

dmitrii-beliakov commented on src-gui/src/main/java/org/openkilda/service/YFlowService.java:
```diff
@@ -0,0 +1,44 @@
+     * @return flow path
+     */
+    public YFlowRerouteResult rerouteFlow(String yFlowId) {
+        LOGGER.info(String.format("Re-rote request for y-flow is in progress, for y-flow id: %s", yFlowId));
```
```suggestion
        LOGGER.info(String.format("The re-routing request for y-flow is in progress, y-flow ID: %s", yFlowId));
```
to reroute (or to re-route) is a verb. The meaning here became different: the rerouting is in progress or the rerouting of the request is in progress. Alternatively you can highlight somehow that `reroute` is the label/name of the request and not the verb in this sentence, for example: REROUTE request.

IvanChupin replied:
fixed

dmitrii-beliakov reviewed (approved):
Great job!
just some minor comments, still would be good to resolve.

yuliiamir commented:
One of the sub-flows has not been reroute during y-flow reroute. 
STR: 
1. Create y-flow that has **shared path** between sub-flows
```
{
  "y_flow_id" :"yflow-sharedPath",
  "shared_endpoint" : {
    "switch_id" : "00:00:00:00:00:00:00:03",
    "port_number" : 11
  },
  "maximum_bandwidth" : 1000,
  "path_computation_strategy" : "COST",
  "encapsulation_type" : "TRANSIT_VLAN",
  "max_latency" : null,
  "max_latency_tier2" : null,
  "ignore_bandwidth" : false,
  "periodic_pings" : false,
  "pinned" : false,
  "priority" : null,
  "strict_bandwidth" : false,
  "description" : "autotest y-flow: All the world 's a stage, and all the men and women merely players. They have their exits and their entrances; And one man in his time plays many parts.",
  "allocate_protected_path" : false,
  "diverse_flow_id" : null,
  "sub_flows" : [ {
    "flow_id" :"subflow1-sharedPath",
    "endpoint" : {
      "switch_id" : "00:00:00:00:00:00:00:02",
      "port_number" : 10,
      "vlan_id" : 221,
      "inner_vlan_id" : 335
    },
    "shared_endpoint" : {
      "vlan_id" : 125,
      "inner_vlan_id" : 112
    },
    "description" : null
  }, {
    "flow_id" : "subflow2-sharedPath",
    "endpoint" : {
      "switch_id" : "00:00:00:00:00:00:00:01",
      "port_number" : 10,
      "vlan_id" : 228,
      "inner_vlan_id" : 448
    },
    "shared_endpoint" : {
      "vlan_id" : 129,
      "inner_vlan_id" : 110
    },
    "description" : null
  } 
 ]
}
```
2. Increase cost for the subflow2-sharedPath segment that is not intersected with subflow1-sharedPath
 subflow1-sharedPath (sw3-sw2) subflow2-sharedPath(sw3-**sw2-sw1)**.
3. Send request to reroute y-flow.
Actual result: 
```
{
    "correlation_id": "30b1dfba-f60c-4259-b92a-01ca9850d80e : 85f5976d-0184-4b92-876f-d475b5227db7",
    "timestamp": 1695719628823,
    "error-type": "The flow already on the best path",
    "error-message": "Could not reroute y-flow",
    "error-description": "Reroute is unsuccessful. Couldn't find new path(s)"
}
```

Expected result: 
subflow2-sharedPath has been reroute.

IvanChupin commented:
> One of the sub-flows has not been reroute during y-flow reroute. STR:
> 
> 1. Create y-flow that has **shared path** between sub-flows
> 
> ```
> {
>   "y_flow_id" :"yflow-sharedPath",
>   "shared_endpoint" : {
>     "switch_id" : "00:00:00:00:00:00:00:03",
>     "port_number" : 11
>   },
>   "maximum_bandwidth" : 1000,
>   "path_computation_strategy" : "COST",
>   "encapsulation_type" : "TRANSIT_VLAN",
>   "max_latency" : null,
>   "max_latency_tier2" : null,
>   "ignore_bandwidth" : false,
>   "periodic_pings" : false,
>   "pinned" : false,
>   "priority" : null,
>   "strict_bandwidth" : false,
>   "description" : "autotest y-flow: All the world 's a stage, and all the men and women merely players. They have their exits and their entrances; And one man in his time plays many parts.",
>   "allocate_protected_path" : false,
>   "diverse_flow_id" : null,
>   "sub_flows" : [ {
>     "flow_id" :"subflow1-sharedPath",
>     "endpoint" : {
>       "switch_id" : "00:00:00:00:00:00:00:02",
>       "port_number" : 10,
>       "vlan_id" : 221,
>       "inner_vlan_id" : 335
>     },
>     "shared_endpoint" : {
>       "vlan_id" : 125,
>       "inner_vlan_id" : 112
>     },
>     "description" : null
>   }, {
>     "flow_id" : "subflow2-sharedPath",
>     "endpoint" : {
>       "switch_id" : "00:00:00:00:00:00:00:01",
>       "port_number" : 10,
>       "vlan_id" : 228,
>       "inner_vlan_id" : 448
>     },
>     "shared_endpoint" : {
>       "vlan_id" : 129,
>       "inner_vlan_id" : 110
>     },
>     "description" : null
>   } 
>  ]
> }
> ```
> 
> 2. Increase cost for the subflow2-sharedPath segment that is not intersected with subflow1-sharedPath
>    subflow1-sharedPath (sw3-sw2) subflow2-sharedPath(sw3-**sw2-sw1)**.
> 3. Send request to reroute y-flow.
>    Actual result:
> 
> ```
> {
>     "correlation_id": "30b1dfba-f60c-4259-b92a-01ca9850d80e : 85f5976d-0184-4b92-876f-d475b5227db7",
>     "timestamp": 1695719628823,
>     "error-type": "The flow already on the best path",
>     "error-message": "Could not reroute y-flow",
>     "error-description": "Reroute is unsuccessful. Couldn't find new path(s)"
> }
> ```
> 
> Expected result: subflow2-sharedPath has been reroute.

Fixed!
the cases that you have described:
changed the costs between sw2-sw1  (700->70000) 
reroute is working for both affinity groups: subflow1-sharedPath and  affinity group is subflow2-sharedPath.
